### ISSUE TITLE
Why is "moduleResolution" an ignored runtime option for Typescript?

### ISSUE THREAD

**Issue by kevlarr** [denoland/deno#10742]
"moduleResolution" is [listed as an ignored](https://github.com/denoland/deno/blob/main/cli/config_file.rs#L115) compiler option, but why is this necessary? Ie. why does it fall under...

> there are lots of parts of a TypeScript configuration file that are meaningless in a Deno context or would cause Deno to not function properly if they were applied.

... from https://deno.land/manual/typescript/configuration#how-deno-uses-a-configuration-file

`compilerOptions.moduleResolution` is **meaningful** (as it relates to loading third-party code) but I fail to see how it causes Deno to function improperly.

Typescript itself says that the "classic" strategy (which is what Deno uses) [should only exist for backwards-compatability](https://www.typescriptlang.org/docs/handbook/module-resolution.html#classic) which means that a lot of projects written in Typescript are bound to use Node's strategy where importing a directory will look (by default) for an `index.ts` file.

**Comment by nayeemrmn** [comment]
Deno does not use the "classic" strategy described in https://www.typescriptlang.org/docs/handbook/module-resolution.html#classic. It uses browser-like ESM module resolution, which doesn't involve guessing and trying to fetch multiple URLs for one module specifier.

**Comment by kevlarr** [comment]
@nayeemrmn Okay, ignore my use of "classic" term; the rest of my comment still stands. Typescript encourages "node" module strategy, so why disallow that **as a compiler option**?

**Comment by lucacasonato** [comment]
Because it is fundamentally incompatible with Deno. Module resolution in Deno aligns to how it works in the browser, not to Node or TSC.

**Comment by kevlarr** [comment]
So it's an ideological reason rather than a technical one? That automatically excludes us from using a lot of Typescript code.

**Comment by kevlarr** [comment]
I get aligning with the browser, since that's standards-based. But we're dealing with back-end code, unfortunately.

**Comment by nayeemrmn** [comment]
Nope, preventing the user from opting back into tsc module resolution is entirely technical.

**Comment by kevlarr** [comment]
@nayeemrmn No. You're talking about a technical "solution" to an ideological desire, because it's about **not wanting** to support node-style imports; it's not about **not being able** to.

**Comment by kevlarr** [comment]
If the broad community wants node-style imports, why **explicitly decide not to allow it** when it's already technically possible, implemented, and tested?

**Comment by nayeemrmn** [comment]
Just FYI, the Node module resolution algorithm described in https://www.typescriptlang.org/docs/handbook/module-resolution.html#how-nodejs-resolves-modules which TS's outdated and unfortunate resolution tries to mimic is something Node **explicitly decided not to allow** in its ESM loader. Try it. The broad community is actively trying to move away from this. Even if it is strictly an ideological desire which again it isn't.

**Comment by lucacasonato** [comment]
> No. You're talking about a technical "solution" to an ideological desire, because it's about **not wanting** (ideological problem) to support node-style imports, rather than about **not being able** (technical problem) to.

No. This is not true. Our infrastructure (remote loading of modules from the web) is **NOT** capable of node resolution. It is technically incompatible. It is not technically possible with our infrastructure, so we do not allow it.

If you dislike this, then maybe Deno is not the right tool for you.

**Comment by lucacasonato** [comment]
> when it's already technically possible, implemented, and tested

It is not. Our TS host implementation (the thing typescript uses to get files from disk), is not capable of node style resolving. Enabling it would not "just work".

**Comment by kevlarr** [comment]
> No. This is not true. Our infrastructure (remote loading of modules from the web) is NOT capable of node resolution. It is technically incompatible. It is not technically possible with our infrastructure, so we do not allow it.

See, this is a technical reason, and it's what I was asking for in the first place. Thank you for explaining.

> If you dislike this, then maybe Deno is not the right tool for you.

There was never a mention that it wasn't technically possible, just that it was **an ignored compiler option**. I asked why, and it took a few rounds of back and forth. It was not about "dislike", it was about asking why the Typescript option was disabled. You could have simply responded with all of this in the first place.

**Comment by lucacasonato** [comment]
> There was never a mention that it wasn't technically possible

I did say "it is fundamentally incompatible with Deno" in the 3rd comment on this PR. I think that is a pretty technical reason.

**Comment by kevlarr** [comment]
@nayeemrmn Thanks for clarifying. The earlier comments made no allusion to technical reasons, so there was no reason for me to assume there *was* a technical reason.

**Comment by kevlarr** [comment]
@lucacasonato "fundamentally" does not mean the same thing as "technically". I shouldn't get too... technical... (eh? eh?) with meaning, though, so I'll just drop it.

**Comment by nayeemrmn** [comment]
> so there was no reason for me to assume there _was_ a technical reason.

Did you read the sentence you quoted from the manual?

**Comment by kevlarr** [comment]
I honestly don't understand why you both have the tone you have had. My post wasn't hostile, it didn't make any value judgement on a feature, it didn't express any dislike - it simply expressed confusion on something I thought should have been possible (after source diving for an hour) and simply asked "why" regarding something that isn't explained in the manual.

Yes, I recognize that the manual can't record a technical reason why every single compiler option for Typescript isn't supported; it would no longer be a useful manual if it had to state every single design decision, because it would be thousands of pages long and be impossible for newcomers to parse. That is why I asked here, because (1) I wanted to know if it could be implemented and (2) it at least exposes it for other people to find so they don't ask the same thing.

I don't personally understand why those points were so hard to achieve.

You are involved in open-source software; you should be able to handle someone asking "why" without being dismissive or rude.

**Comment by nayeemrmn** [comment]
You repeatedly made assertions (not inquiries) about how the hurdle isn't technical and that it should be easy, _after_ having been told otherwise. The question was "why don't you do it, considering this?".

**Comment by kevlarr** [comment]
Okay, I finally see the problem. Huge misunderstanding.

I asked "**Why** is option X ignored..." and I wanted to know **the** reason *if* there was one; I didn't ask "**Is** there a technical reason...".

Then, your statements didn't provide context other than asserting it's technical (and you _wouldn't_ provide context if it was a simple"is there a reason.." question, so that makes sense) but without your knowledge I read...

> Nope, preventing the user from opting back into tsc module resolution is entirely technical"

... more as meaning something like...

>  we *decided* to not let users opt back in, and that decision itself *was* the technical reason

... primarily because of @lucacasonato comment about "aligning" which, like I alluded to, is a way people describe taking ideological stances.

I then made assertions about *your responses themselves* based on my reading of your comment (since I don't have your knowledge underlying all of your comments), and you then took that as me making assertions *about the hurdles* - and only after those exchanges did either of you describe the technical reasons and, fittingly, closed the issue.

I also made assertions about "being implemented" and I was referring to "module resolution" stuff in the the bundled Typescript compiler, but I wasn't clear in that, so my bad.

I'll try to be more explicit about questions elsewhere in the future, so this kind of thing isn't duplicated, and if someone else's comments don't make sense in the context of me asking, I'll ask to clarify rather than making assumptions. I would definitely suggest doing the same and avoiding inflammatory comments like "Did you even read..." in the future. (Which was especially insulting after I'd already thanked you for clarifying.)

**tldr** I misunderstood both of yours comments and both of you seemingly misunderstood mine, and none of us sought clarification, and here we are.
